File: extjvm/attachment/remote_jvm_connections/remote_jvm_connections.go
package remote_jvm_connections

import (
	"github.com/rs/zerolog/log"
	"github.com/steadybit/extension-jvm/extjvm/utils"
	"strconv"
	"sync"
	"time"
)

type SocketConnection struct {
	Host  string
	Port  int
	Mutex sync.Mutex
}

func (a *SocketConnection) Address() string {
	return a.Host + ":" + strconv.Itoa(a.Port)
}

var (
	connections = sync.Map{} //map[int32]SocketConnection (IP address)
)

func WaitForConnection(pid int32, timeout time.Duration) bool {
	wg := sync.WaitGroup{}
	wg.Add(1)
	go func() {
		for {
			if GetConnection(pid) != nil {
				wg.Done()
				return
			}
			time.Sleep(100 * time.Millisecond)
		}
	}()
	if utils.WaitTimeout(&wg, timeout) {
		return false
	} else {
		return true
	}
}

func GetConnection(pid int32) *SocketConnection {
	ipaddress, ok := connections.Load(pid)
	if !ok {
		log.Trace().Msgf("No connection found for pid %d", pid)
		return nil
	}
	return ipaddress.(*SocketConnection)
}

func AddConnection(pid int32, host string, port int) {
	log.Info().Msgf("Adding connection for PID %d on %s:%d", pid, host, port)
	connection := SocketConnection{Host: host, Port: port}
	existingConnection := GetConnection(pid)
	if existingConnection != nil && existingConnection.Host == host && existingConnection.Port == port {
		log.Debug().Msgf("JVM connection with with PID %d on %s exists already. Skipping registration", pid, connection.Address())
		return
	}
	connectionIsNew := false
	if existingConnection == nil {
		connectionIsNew = true
	}
	connections.Store(pid, &connection)
	if connectionIsNew {
		log.Debug().Msgf("JVM connection with PID %d on %s registered", pid, connection.Address())
	} else {
		log.Debug().Msgf("JVM connection with PID %d on %s updated", pid, connection.Address())
	}
}

func RemoveConnection(pid int32) {
	connections.Delete(pid)
	log.Trace().Msgf("JVM connection with PID %d removed", pid)
}

func ClearConnections() {
	connections = sync.Map{}
	log.Debug().Msg("All JVM connections removed")
}

func Size() int {
	size := 0
	connections.Range(func(key, value interface{}) bool {
		size++
		return true
	})
	return size
}

File: javaagents/agent-bundles-core/javaagent-setup/src/main/java/com/steadybit/javaagent/attachment/internal/RemoteJvmConnections.java
/*
 * Copyright 2022 steadybit GmbH. All rights reserved.
 */

package com.steadybit.javaagent.attachment.internal;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class RemoteJvmConnections {
    private static final Logger log = LoggerFactory.getLogger(RemoteJvmConnections.class);
    private final Map<Integer, InetSocketAddress> connections = new ConcurrentHashMap<>();

    public boolean waitForConnection(Integer pid, long timeout) {
        var deadline = System.currentTimeMillis() + timeout;
        try {
            while (System.currentTimeMillis() <= deadline) {
                if (this.connections.containsKey(pid)) {
                    return true;
                }
                Thread.sleep(100L);
            }
        } catch (InterruptedException e) {
            log.warn("Interrupted waiting for RemoteJvmConnection.");
            //interrupt self so that the attachment loop in javadiscovery for multiple JVMs  is stopped
            Thread.currentThread().interrupt();
        }
        return false;
    }

    public InetSocketAddress getConnection(Integer pid) {
        return this.connections.get(pid);
    }

    public void addConnection(Integer pid, InetAddress host, Integer port) {
        var connection = new InetSocketAddress(host, port);

        var existingConnection = this.connections.get(pid);
        if (existingConnection != null && existingConnection.equals(connection)) {
            log.trace("JVM connection with  with PID {} on {} exists already. Skipping registration.", pid, connection);
            return;
        }

        existingConnection = this.connections.put(pid, connection);
        if (existingConnection == null) {
            log.debug("Registered new JVM connection with PID {} on {}", pid, connection);
        } else {
            log.debug("Replaced existing JVM connection with PID {} on {}", pid, connection);
        }
    }

    public void removeConnection(Integer pid) {
        this.connections.remove(pid);
    }

    public Integer size() {
        return this.connections.size();
    }

    public void clear() {
        this.connections.clear();
    }
}
